numpy/_core/src/multiarray/getset.c

```diff
@@ -848,6 +848,18 @@ array_flat_set(PyArrayObject *self, PyObject *val, void *NPY_UNUSED(ignored))
 static PyObject *
 array_transpose_get(PyArrayObject *self, void *NPY_UNUSED(ignored))
 {
+    int ndim = PyArray_NDIM(self);
+    if (ndim != 2) {
+        if (PyErr_WarnFormat(PyExc_UserWarning, 1,
```

Comment:
Why a `UserWarning`?  Please use a `DeprecationWarning`, or actually, use the `DEPRECATE` macro.

We could use a visible deprecation warning if we want to inform new users who accidentally get it wrong. But only if we take the trouble to check that very few libraries use this (i.e. there are few places in existing code where `DeprecationWarning` would be hidden, but `VisibleDeprecationWarning` wouldn't be).

Reply:
> Given the number of places the tests needed to be updated, I bet this will be a noisy deprecation.

So, it must clearly be a `DeprecationWarning`.  This should be taken very slowly, i.e. in a year or so a `VisibleDeprecationWarning`.

Reply:
Done - for array scalars in `scalartypes.c.src` I used DEPRECATE macro, but here I kept `PyErr_WarnFormat`, now with a `DeprecationWarning`, because it allows formatting contrary to `DEPRECATE`.